tls/s2n_early_data.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_early_data.h"
+
+#include "tls/s2n_connection.h"
+#include "utils/s2n_safety.h"
+
+const s2n_early_data_state valid_previous_states[] = {
+        [S2N_EARLY_DATA_REQUESTED]      = S2N_UNKNOWN_EARLY_DATA_STATE,
+        [S2N_EARLY_DATA_NOT_REQUESTED]  = S2N_UNKNOWN_EARLY_DATA_STATE,
+        [S2N_EARLY_DATA_REJECTED]       = S2N_EARLY_DATA_REQUESTED,
+        [S2N_EARLY_DATA_ACCEPTED]       = S2N_EARLY_DATA_REQUESTED,
+        [S2N_END_OF_EARLY_DATA]         = S2N_EARLY_DATA_ACCEPTED,
+};
+
+S2N_RESULT s2n_connection_set_early_data_state(struct s2n_connection *conn, s2n_early_data_state next_state)
+{
+    ENSURE_REF(conn);
+    ENSURE(next_state < S2N_EARLY_DATA_STATES_COUNT, S2N_ERR_INVALID_EARLY_DATA_STATE);
+    ENSURE(next_state != S2N_UNKNOWN_EARLY_DATA_STATE, S2N_ERR_INVALID_EARLY_DATA_STATE);
+    ENSURE(conn->early_data_state == valid_previous_states[next_state], S2N_ERR_INVALID_EARLY_DATA_STATE);
```

Comment:
Q: Is there a reason behind being specific with the safety error code here and using `S2N_ERR_INVALID_EARLY_DATA_STATE ` instead of `S2N_ERR_SAFETY` as we do in other places in s2n?
We could just use `ENSURE_LT`, `ENSURE_NE` and `ENSURE_EQ`.

Reply:
Having a specific error rather than S2N_ERR_SAFETY will make debugging any issue involving the early data state much easier. I'm thinking about similar features like the certificate validation state where the unique error code really helps.

Reply:
It's mostly a tradeoff of generic/specific errors. `S2N_ERR_SAFETY` doesn't actually tell you a whole lot about why the call failed or what it was expected, whereas using a specific error like this one gives more information to the caller.